enterprise/server/remote_execution/copy_on_write/copy_on_write.go

```diff
@@ -689,19 +691,56 @@ func ConvertFileToCOW(ctx context.Context, env environment.Env, filePath string,
 
 	// Copy buffer (large enough to copy one data block at a time).
 	ioBlockSize := int64(stat.Sys().(*syscall.Stat_t).Blksize)
-	copyBuf := make([]byte, ioBlockSize)
+	copyBufPool := sync.Pool{
+		New: func() any {
+			copyBuf := make([]byte, ioBlockSize)
+			return &copyBuf
+		},
+	}
+
+	// Seeking moves the offset of the file descriptor, so lock to ensure concurrent
+	// goroutines are simultaneously moving it
+	var fileMu sync.Mutex
+
+	// readFromOffset starts reading from `startOffset`, but will skip forward if there
+	// is no data at `startOffset`.
+	// Returns the offset it actually started reading from
+	// (the offset of the first non-empty data block >= `startOffset`)
+	readFromOffset := func(startOffset int64, buf []byte) (int64, error) {
+		fileMu.Lock()
+		defer fileMu.Unlock()
+
+		// Seek to the start of a non-empty data block in the file
+		// greater or equal to `startOffset`
+		dataOffset, err := syscall.Seek(fd, startOffset, unix.SEEK_DATA)
+		if err == syscall.ENXIO {
+			// No more data in the file
+			return totalSizeBytes, nil
+		} else if err != nil {
+			return 0, err
+		}
 
-	// dataOffset points to the start of the current non-empty data block in the
-	// file. It is updated every time we consume a data block.
-	dataOffset, err := syscall.Seek(fd, 0, unix.SEEK_DATA)
-	if err == syscall.ENXIO {
-		// No more data
-		dataOffset = totalSizeBytes
-	} else if err != nil {
-		return nil, status.InternalErrorf("initial data seek failed: %s", err)
+		// TODO: see whether it's faster to mmap the file we're copying
+		// from, rather than issuing read() syscalls for each data block
+		if _, err := io.ReadFull(f, buf); err != nil {
+			return 0, err
+		}
+		return dataOffset, nil
 	}
 
 	createChunk := func(chunkStartOffset int64) (*Mmap, error) {
+		// Check if there's any data in the chunk, to avoid writing a file if
+		// there isn't
+		fileMu.Lock()
+		dataOffset, err := syscall.Seek(fd, chunkStartOffset, unix.SEEK_DATA)
+		fileMu.Unlock()
```

Comment:
we could avoid the file mutex by having one open file handle per goroutine, so that they can be `seek`ed independently. I'm not sure if that will improve performance, but might simplify things. Maybe worth trying out?

Reply:
Another option is to have `readFromOffset` only lock around the `syscall.Seek`, and then you could use the `readFullAt` func in this package instead of `io.ReadFull`.